cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -0,0 +1,134 @@
+  private static final Logger LOG = LoggerFactory.getLogger(InMemorySourceControlOperationRunner.class);
+
+  @Inject
+  public InMemorySourceControlOperationRunner(RepositoryManagerFactory repoManagerFactory) {
+    this.repoManagerFactory = repoManagerFactory;
+  }
+
+  @Override
+  public PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
+                              ApplicationDetail appToPush, CommitMeta commitDetails)
+    throws Exception {
+    try (RepositoryManager repositoryManager = repoManagerFactory.create(namespace, repoConfig)) {
+      try {
+        repositoryManager.cloneRemote();
+      } catch (Exception e) {
+        throw new PushFailureException("Failed to clone remote repository", e);
+      }
+      LOG.info("Pushing application configs for : {}", appToPush);
+
+      //TODO CDAP-20371: Add retry logic here in case the head at remote moved while we are doing push
+      String pushedCommit = writeAppDetailAndPush(repositoryManager, appToPush, commitDetails);
+
+      try {
+        Path relativeFilePath = getAndValidateAppConfigRelativePath(repositoryManager, appToPush.getName());
+        String gitFileHash = repositoryManager.getFileHash(relativeFilePath, pushedCommit);
+        return new PushAppResponse(appToPush.getName(), appToPush.getAppVersion(), gitFileHash);
+      } catch (Exception e) {
+        throw new PushFailureException("Failed to fetch push details", e);
+      }
+    }
+  }
+
+  private String writeAppDetailAndPush(RepositoryManager repositoryManager,
+                                       ApplicationDetail appToPush,
+                                       CommitMeta commitDetails)
+    throws PushFailureException, NoChangesToPushException {
+    Path appWrittenPath;
+    try {
+      String appName = appToPush.getName();
+      // Symlinks will be followed but if the path points to an outside directory it will fail
+      appWrittenPath = getAndValidateAppConfigRelativePath(repositoryManager, appName);
+      // Opens the file for writing, creating the file if it doesn't exist,
+      // or truncating an existing regular-file to a size of 0
+      Path filePathToWrite = repositoryManager.getRepositoryRoot().resolve(appWrittenPath);
+      try (FileWriter writer = new FileWriter(filePathToWrite.toString())) {
+        GSON.toJson(appToPush, writer);
+      }
+
+      LOG.debug("Written application configs for {} in file {}", appName, appWrittenPath);
+    } catch (Exception e) {
+      throw new PushFailureException("Failed to write app configs", e);
+    }
+
+    try {
+      String pushedCommit = repositoryManager.commitAndPush(commitDetails, Collections.singletonList(appWrittenPath));
+      LOG.info("Pushed application configs for : {}", appToPush);
+      return pushedCommit;
+    } catch (UnexpectedRepositoryChangesException | GitAPIException e) {
+      throw new PushFailureException("Failed to push to git", e);
```

Comment:
Should be very careful whenever we do something like this. Are we sure the caller never wants to differentiate between these situations? They seem pretty different.

Reply:
Updated the exception handling, also added a TODO: https://cdap.atlassian.net/browse/CDAP-20401
@samdgupi

Reply:
adding a todo is a great way to lose context and not do the proper thing later. Spend a little time now to figure out what is the right thing vs more time later.

For example, getFileHash() throws NotFoundException if the file isn't there, but how is that even possible if we just committed it and pushed? This is one of the reasons it doesn't make sense to even have that method, but it should be coming back in the commitAndPush() method.

getFileHash() also throws IllegalArgumentException if the path is not a file (which shouldn't be possible), but this method throws InvalidPathInSourceControlException. It will be an total mess trying to use this properly without fixing it now.

Reply:
As we discussed offline, instead of calling `getFileHash()`, we just get the fileHash from `commitAndPush`.